### PR TITLE
Feature/oauth2

### PR DESCRIPTION
**Pull request by alahtarin**

### CODE REVIEW

File: lib/Command/CheckoutSession.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+class CheckoutSession
+{
+    /**
+     * @var string
+     */
+    public $merchantCustomerId;
+
+    /**
+     * RetrieveOrder constructor.
```
Comment (alahtarin) [inline lib/Command/CheckoutSession.php]: wrong comment

File: lib/Command/CheckoutSession.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+     *
+     * @param string $merchantCustomerId
+     */
+    public function __construct($merchant_customer_id)
```
Comment (alahtarin) [inline lib/Command/CheckoutSession.php]: the code style rules suggest $camelCase

File: lib/Command/CheckoutSession.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * @param ClassMetadata $metadata
+     */
+    public static function loadValidatorMetadata(ClassMetadata $metadata)
+    {
+        $metadata->addPropertyConstraints('merchantCustomerId', [
+            new Assert\NotBlank()
+        ]);
+    }
+}
```
Comment (alahtarin) [inline lib/Command/CheckoutSession.php]: line break should be added at the end of the file

File: lib/Command/CheckoutSessionConfirm.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * @package Billie\Command
+ * @author Niels Gongoll
+ */
+class CheckoutSessionConfirm
+{
+    /**
+     * @var string
+     */
+    public $uuid;
+    /**
```
Comment (alahtarin) [inline lib/Command/CheckoutSessionConfirm.php:23]: please add empty lines between properties

File: lib/Command/PreapproveCreateOrder.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * @var Amount
+     */
+    public $amount;
+    /**
+     * @var int
+     *
+     * Number of days, the debtor has to pay the invoice (e.g. 14)
+     */
+    public $duration = 14;
```
Comment (alahtarin) [inline lib/Command/PreapproveCreateOrder.php]: why do we need default value? it should be required and always provided

File: lib/HttpClient/BillieClient.php [hunk 6]
```diff
@@ -64,24 +71,90 @@ public function __construct(ValidatorInterface $validator)
...
+
+        try {
+
+            // Try to get an access token using the client credentials grant.
+            $accessToken = $provider->getAccessToken('client_credentials');
+
+        } catch (\League\OAuth2\Client\Provider\Exception\IdentityProviderException $e) {
+
+            // Failed to get the access token
+            exit($e->getMessage());
```
Comment (alahtarin) [inline lib/HttpClient/BillieClient.php]: why exit? what about throwing an exception?

File: lib/HttpClient/BillieClient.php [hunk 7]
```diff
@@ -64,24 +71,90 @@ public function __construct(ValidatorInterface $validator)
...
+        return $result['id'];
+
+    }
+
+    public function checkoutSessionConfirm(CheckoutSessionConfirm $checkoutSessionConfirm)
+    {
+
+        // validate input
+        if ($violations = $this->validateCommand($checkoutSessionConfirm)) {
+            print_r($violations);
```
Comment (alahtarin) [inline lib/HttpClient/BillieClient.php]: not needed

File: lib/HttpClient/BillieClient.php [hunk 8]
```diff
@@ -102,6 +175,83 @@ public function getOrder($orderId)
...
+            }
+        }
+
+        // if houseNumber is empty, set fullAddress to trigger full-address-recognition
+        if (!isset($preapproveCreateOrderCommand->deliveryAddress->fullAddress)
+            && empty($preapproveCreateOrderCommand->deliveryAddress->houseNumber)) {
+            $preapproveCreateOrderCommand->deliveryAddress->fullAddress = $preapproveCreateOrderCommand->deliveryAddress->street;
+        }
+
+        if (isset($preapproveCreateOrderCommand->deliveryAddress->fullAddress)) {
```
Comment (alahtarin) [inline lib/HttpClient/BillieClient.php:189]: these two blocks of the code look very similar, can they be refactored by creating a separate method and calling it twice with diifferent data?

File: lib/HttpClient/BillieClient.php [hunk 9]
```diff
@@ -102,6 +175,83 @@ public function getOrder($orderId)
...
+
+    public function preapproveConfirmOrder(PreapproveConfirmOrder $command)
+    {
+
+        // validate input
+        if ($violations = $this->validateCommand($command)) {
+            throw new InvalidCommandException($violations);
+        }
+
+        $data = array();
```
Comment (alahtarin) [inline lib/HttpClient/BillieClient.php:230]: not needed, you can just pass `array()` to the request method

File: lib/HttpClient/BillieClient.php [hunk 10]
```diff
@@ -308,8 +458,10 @@ private function getClient()
                     'base_url' => $this->apiBaseUrl,
                     'defaults' => [
                         'headers' => [
-                            'X-API-KEY'    => $this->apiKey,
+//                            'X-API-KEY'    => $this->apiKey,
```
Comment (alahtarin) [inline lib/HttpClient/BillieClient.php]: please remove

File: vendor/composer/semver/CHANGELOG.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
+# Change Log
```
Comment (alahtarin) [inline vendor/composer/semver/CHANGELOG.md]: why is the vendor folder added to git?

Comment (ngongoll) [reply vendor/composer/semver/CHANGELOG.md]: magento 1 is not supporting composer from scratch. So we compiled a version and added it to the repo. if M1 is not longer supported we can remove this folder.

File: assets/billie_bankaccount.json [hunk 12]
```diff
@@ -0,0 +1,6 @@
+[
```
Comment (rommelfreddy) [inline assets/billie_bankaccount.json:1]: Hey @ngongoll warum ist hier nur die eine Bank drin? ich hab neulich eine CSV von euch erhalten mit allen Banken.
Wenn die Daten aus der CSV auch im SDK wären, dann könnten wir das json nutzen.

**Pull request closed without merging**